In [1]:
import pandas as pd
import numpy as np

###### I've tried to download the file and I've found it is too heavy to download it directly, so lo lets download first 10 rows

In [2]:
bank = pd.read_csv('./data/train_ver2.csv', nrows=10)
bank.head()

fecha_dato  ncodpers ind_empleado pais_residencia sexo  age  fecha_alta  \
0  2015-01-28   1375586            N              ES    H   35  2015-01-12   
1  2015-01-28   1050611            N              ES    V   23  2012-08-10   
2  2015-01-28   1050612            N              ES    V   23  2012-08-10   
3  2015-01-28   1050613            N              ES    H   22  2012-08-10   
4  2015-01-28   1050614            N              ES    V   23  2012-08-10   

   ind_nuevo  antiguedad  indrel  ...  ind_hip_fin_ult1  ind_plan_fin_ult1  \
0          0           6       1  ...                 0                  0   
1          0          35       1  ...                 0                  0   
2          0          35       1  ...                 0                  0   
3          0          35       1  ...                 0                  0   
4          0          35       1  ...                 0                  0   

  ind_pres_fin_ult1 ind_reca_fin_ult1 ind_tjcr_fin_ult1  ind_valo_fin_ult1  \
0                 0                 0                 0                  0   
1                 0                 0                 0                  0   
2                 0                 0                 0                  0   
3                 0                 0                 0                  0   
4                 0                 0                 0                  0   

  ind_viv_fin_ult1 ind_nomina_ult1  ind_nom_pens_ult1  ind_recibo_ult1  
0                0               0                  0                0  
1                0               0                  0                0  
2                0               0                  0                0  
3                0               0                  0                0  
4                0               0                  0                0  

[5 rows x 48 columns]

In [3]:
bank.columns

Index(['fecha_dato', 'ncodpers', 'ind_empleado', 'pais_residencia', 'sexo',
       'age', 'fecha_alta', 'ind_nuevo', 'antiguedad', 'indrel',
       'ult_fec_cli_1t', 'indrel_1mes', 'tiprel_1mes', 'indresi', 'indext',
       'conyuemp', 'canal_entrada', 'indfall', 'tipodom', 'cod_prov',
       'nomprov', 'ind_actividad_cliente', 'renta', 'segmento',
       'ind_ahor_fin_ult1', 'ind_aval_fin_ult1', 'ind_cco_fin_ult1',
       'ind_cder_fin_ult1', 'ind_cno_fin_ult1', 'ind_ctju_fin_ult1',
       'ind_ctma_fin_ult1', 'ind_ctop_fin_ult1', 'ind_ctpp_fin_ult1',
       'ind_deco_fin_ult1', 'ind_deme_fin_ult1', 'ind_dela_fin_ult1',
       'ind_ecue_fin_ult1', 'ind_fond_fin_ult1', 'ind_hip_fin_ult1',
       'ind_plan_fin_ult1', 'ind_pres_fin_ult1', 'ind_reca_fin_ult1',
       'ind_tjcr_fin_ult1', 'ind_valo_fin_ult1', 'ind_viv_fin_ult1',
       'ind_nomina_ult1', 'ind_nom_pens_ult1', 'ind_recibo_ult1'],
      dtype='object')

In [ ]:
# I've used this to get the dataframe by column and, with the three cells below, determine the optimal cell type.
# The purpose of doing that is open all dataframe for feature enginering.

usecols = ['ind_pres_fin_ult1', 'ind_reca_fin_ult1',
       'ind_tjcr_fin_ult1', 'ind_valo_fin_ult1', 'ind_viv_fin_ult1',
       'ind_nomina_ult1', 'ind_nom_pens_ult1', 'ind_recibo_ult1']

bank_by_cols = pd.read_csv('./data/train_ver2.csv', usecols=usecols
                         , na_values={'na':0, 'NA':0 , 'NaN':0.0, 'nan':0.0}
                        )


In [ ]:
# This helps to reduce the bytes of integers to minimum (int8) if possible

for int_col in bank_total.select_dtypes('integer').columns:
    bank_total[int_col] = pd.to_numeric(bank_total[int_col], downcast='integer')
    
bank_total.info(memory_usage='deep')

In [ ]:
# same for float

for int_col in bank_total.select_dtypes('float').columns:
    bank_total[int_col] = pd.to_numeric(bank_total[int_col], downcast='float')
    
bank_total.info(memory_usage='deep')

In [ ]:
# Convert object to category - Will be used later

for cat_feature in bank_total.select_dtypes('object').columns:
    bank_total[cat_feature] = bank_total[cat_feature].astype('category')

bank_total.info(memory_usage='deep')

In [4]:
# Dictionary with all columns and its types, this step is made in order to download the full file

cols_types = {'fecha_dato':'object',
'ncodpers':'int32',
'ind_empleado':'object',
'pais_residencia':'object',
'sexo':'object',
'age':'object',          # este tiene un valor NA, en la doc de pandas se puede decir NA=0
'fecha_alta':'object',
'ind_nuevo':'float32',
'antiguedad':'object',   # este tiene un valor NA, en la doc de pandas se puede decir NA=0
'indrel':'float32',
              
'ult_fec_cli_1t':'object',
'indrel_1mes':'object',
'tiprel_1mes':'object',
'indresi':'object',
'indext':'object',
'conyuemp':'object',
'canal_entrada':'object',
'indfall':'object',
'tipodom':'float32',
'cod_prov':'float32',    # debería ser int32
              
'nomprov':'object',
'ind_actividad_cliente':'float32',
'renta':'float32',
'segmento':'object',
'ind_ahor_fin_ult1':'int8',
'ind_aval_fin_ult1':'int8',
'ind_cco_fin_ult1':'int8',
'ind_cder_fin_ult1':'int8',
'ind_cno_fin_ult1':'int8',
'ind_ctju_fin_ult1':'int8',
              
'ind_ctma_fin_ult1':'int8',
'ind_ctop_fin_ult1':'int8',
'ind_ctpp_fin_ult1':'int8',
'ind_deco_fin_ult1':'int8',
'ind_deme_fin_ult1':'int8',
'ind_dela_fin_ult1':'int8',
'ind_ecue_fin_ult1':'int8',
'ind_fond_fin_ult1':'int8',
'ind_hip_fin_ult1':'int8',
'ind_plan_fin_ult1':'int8',
              
'ind_pres_fin_ult1':'int8',
'ind_reca_fin_ult1':'int8',
'ind_tjcr_fin_ult1':'int8',
'ind_valo_fin_ult1':'int8',
'ind_viv_fin_ult1':'int8',
'ind_nomina_ult1':'float32',
'ind_nom_pens_ult1':'float32',
'ind_recibo_ult1':'int8'
             }

In [6]:
bank_total = pd.read_csv('./data/train_ver2.csv', dtype=cols_types
                         , na_values={'na':0, 'NA':0 , 'NaN':0.0, 'nan':0.0}
                        )
bank_total.head()

fecha_dato  ncodpers ind_empleado pais_residencia sexo  age  fecha_alta  \
0  2015-01-28   1375586            N              ES    H   35  2015-01-12   
1  2015-01-28   1050611            N              ES    V   23  2012-08-10   
2  2015-01-28   1050612            N              ES    V   23  2012-08-10   
3  2015-01-28   1050613            N              ES    H   22  2012-08-10   
4  2015-01-28   1050614            N              ES    V   23  2012-08-10   

   ind_nuevo antiguedad  indrel  ... ind_hip_fin_ult1 ind_plan_fin_ult1  \
0        0.0          6     1.0  ...                0                 0   
1        0.0         35     1.0  ...                0                 0   
2        0.0         35     1.0  ...                0                 0   
3        0.0         35     1.0  ...                0                 0   
4        0.0         35     1.0  ...                0                 0   

  ind_pres_fin_ult1 ind_reca_fin_ult1 ind_tjcr_fin_ult1 ind_valo_fin_ult1  \
0                 0                 0                 0                 0   
1                 0                 0                 0                 0   
2                 0                 0                 0                 0   
3                 0                 0                 0                 0   
4                 0                 0                 0                 0   

  ind_viv_fin_ult1 ind_nomina_ult1  ind_nom_pens_ult1  ind_recibo_ult1  
0                0             0.0                0.0                0  
1                0             0.0                0.0                0  
2                0             0.0                0.0                0  
3                0             0.0                0.0                0  
4                0             0.0                0.0                0  

[5 rows x 48 columns]

<font color=blue>

### Finally I got open the file!!
</font>

Lets check the size of the dataframe


In [7]:
bank_total.shape

(13647309, 48)

Now I want to know which columns are categories.

For this purpose I've create a function that prints the number of unique values and the unique values. I prefer doing it on this way since I'm able to check all columns in one step. It consumes more computational resources but I think is worthly.

In [8]:
all_columns = ['fecha_dato', 'ncodpers', 'ind_empleado', 'pais_residencia', 'sexo',
       'age', 'fecha_alta', 'ind_nuevo', 'antiguedad', 'indrel',
       'ult_fec_cli_1t', 'indrel_1mes', 'tiprel_1mes', 'indresi', 'indext',
       'conyuemp', 'canal_entrada', 'indfall', 'tipodom', 'cod_prov',
       'nomprov', 'ind_actividad_cliente', 'renta', 'segmento',
       'ind_ahor_fin_ult1', 'ind_aval_fin_ult1', 'ind_cco_fin_ult1',
       'ind_cder_fin_ult1', 'ind_cno_fin_ult1', 'ind_ctju_fin_ult1',
       'ind_ctma_fin_ult1', 'ind_ctop_fin_ult1', 'ind_ctpp_fin_ult1',
       'ind_deco_fin_ult1', 'ind_deme_fin_ult1', 'ind_dela_fin_ult1',
       'ind_ecue_fin_ult1', 'ind_fond_fin_ult1', 'ind_hip_fin_ult1',
       'ind_plan_fin_ult1', 'ind_pres_fin_ult1', 'ind_reca_fin_ult1',
       'ind_tjcr_fin_ult1', 'ind_valo_fin_ult1', 'ind_viv_fin_ult1',
       'ind_nomina_ult1', 'ind_nom_pens_ult1', 'ind_recibo_ult1']
for col in all_columns:
    print(col, '- num unique values:', bank_total[col].nunique(), '\n', bank_total[col].unique(), '\n\n')

fecha_dato - num unique values: 17 
 ['2015-01-28' '2015-02-28' '2015-03-28' '2015-04-28' '2015-05-28'
 '2015-06-28' '2015-07-28' '2015-08-28' '2015-09-28' '2015-10-28'
 '2015-11-28' '2015-12-28' '2016-01-28' '2016-02-28' '2016-03-28'
 '2016-04-28' '2016-05-28'] 


ncodpers - num unique values: 956645 
 [1375586 1050611 1050612 ... 1173729 1164094 1550586] 


ind_empleado - num unique values: 5 
 ['N' nan 'A' 'B' 'F' 'S'] 


pais_residencia - num unique values: 118 
 ['ES' nan 'CA' 'CH' 'CL' 'IE' 'AT' 'NL' 'FR' 'GB' 'DE' 'DO' 'BE' 'AR' 'VE'
 'US' 'MX' 'BR' 'IT' 'EC' 'PE' 'CO' 'HN' 'FI' 'SE' 'AL' 'PT' 'MZ' 'CN'
 'TW' 'PL' 'IN' 'CR' 'NI' 'HK' 'AD' 'CZ' 'AE' 'MA' 'GR' 'PR' 'RO' 'IL'
 'RU' 'GT' 'GA' 'NO' 'SN' 'MR' 'UA' 'BG' 'PY' 'EE' 'SV' 'ET' 'CM' 'SA'
 'CI' 'QA' 'LU' 'PA' 'BA' 'BO' 'AU' 'BY' 'KE' 'SG' 'HR' 'MD' 'SK' 'TR'
 'AO' 'CU' 'GQ' 'EG' 'ZA' 'DK' 'UY' 'GE' 'TH' 'DZ' 'LB' 'JP' 'NG' 'PK'
 'TN' 'TG' 'KR' 'GH' 'RS' 'VN' 'PH' 'KW' 'NZ' 'MM' 'KH' 'GI' 'SL' 'GN'
 'GW' 'OM' 'CG' 'LV' 'LT' '

tipodom - num unique values: 1 
 [ 1. nan] 


cod_prov - num unique values: 52 
 [29. 13. 50. 45. 24. 20. 10. 17. 49.  8. 37.  9. 22. 31.  5. 40. 27. 25.
 28.  3. 42. 41. 39.  7. 47. 36. 46. 44. 15. 32. 23. 16. 48. 12. 26.  2.
  6. 30. 11. nan  4. 19. 34. 35. 14. 21. 18. 33. 38. 52. 43.  1. 51.] 


nomprov - num unique values: 52 
 ['MALAGA' 'CIUDAD REAL' 'ZARAGOZA' 'TOLEDO' 'LEON' 'GIPUZKOA' 'CACERES'
 'GIRONA' 'ZAMORA' 'BARCELONA' 'SALAMANCA' 'BURGOS' 'HUESCA' 'NAVARRA'
 'AVILA' 'SEGOVIA' 'LUGO' 'LERIDA' 'MADRID' 'ALICANTE' 'SORIA' 'SEVILLA'
 'CANTABRIA' 'BALEARS, ILLES' 'VALLADOLID' 'PONTEVEDRA' 'VALENCIA'
 'TERUEL' 'CORUÑA, A' 'OURENSE' 'JAEN' 'CUENCA' 'BIZKAIA' 'CASTELLON'
 'RIOJA, LA' 'ALBACETE' 'BADAJOZ' 'MURCIA' 'CADIZ' nan 'ALMERIA'
 'GUADALAJARA' 'PALENCIA' 'PALMAS, LAS' 'CORDOBA' 'HUELVA' 'GRANADA'
 'ASTURIAS' 'SANTA CRUZ DE TENERIFE' 'MELILLA' 'TARRAGONA' 'ALAVA' 'CEUTA'] 


ind_actividad_cliente - num unique values: 2 
 [ 1.  0. nan] 


renta - num unique values: 520988 
 

I've found that field ncodpers (which is the personal code) have 956,645 values, so that means the number of customers represented on this data. We have 17th values in "fecha_dato" field; from January 2015 to May 2016. So as there is only 13,647,309 rows, that means some customers won't have all records.

The columns that have low number of values are the ones listed below, in order to understand the data, lets reseach what means each field and if it needs to be cleaned(NBC):

- ind_empleado ==> NBC - Categorical - A active, B ex employed, F filial, N not employee, P pasive.
- sexo ==> NBC - Categorical - H women, V men.
- ind_nuevo ==> NBC - Non categorical, it indicates with '1' if customer got registered in the last 6 months.
- indrel ==> NBC - Categorical - 1 First/Primary, 99 Primary customer during the month but not at the end of the month
- indrel_1mes ==> NBC - Categorical - 1 (First/Primary customer), 2 (co-owner ), P (Potential), 3 (former primary), 4(former co-owner)
- tiprel_1mes ==> NBC - Categorical - Customer relation type at the beginning of the month, A (active), I (inactive), P (former customer), R (Potential)
- indresi ==> NBC - Categorical - Residence index - Same residence country than the bank - S (Yes) or N (No)
- indext ==> NBC - Categorical - Foreigner index - Birth country different than bank's residence country - S (Yes) or N (No)
- conyuemp ==> NBC - Categorical - S (married with an employee) N (not married with an employee)
- indfall ==> NBC - Categorical - S (customer has died) N (customer alive)
- tipodom ==> To delete - Address type - 1 for primary residence
- ind_actividad_cliente ==> NBC - 1, active customer; 0, inactive customer
- segmento ==> NBC - Kind of customer
- ind_ahor_fin_ult1 ==> Target column - Saving account
- ind_aval_fin_ult1 ==> Target column - Guarantees
- ind_cco_fin_ult1 ==> Target column - Current Accounts
- ind_cder_fin_ult1 ==> Target column - Derivada Account
- ind_cno_fin_ult1 ==> Target column - Payroll Account
- ind_ctju_fin_ult1 ==> Target column - Junior Account
- ind_ctma_fin_ult1 ==> Target column - Más particular Account
- ind_ctop_fin_ult1 ==> Target column - particular Account
- ind_ctpp_fin_ult1 ==> Target column - particular Plus Account
- ind_deco_fin_ult1 ==> Target column - Short-term deposits
- ind_deme_fin_ult1 ==> Target column - Medium-term deposits
- ind_dela_fin_ult1 ==> Target column - Long-term deposits
- ind_ecue_fin_ult1 ==> Target column - e-account
- ind_fond_fin_ult1 ==> Target column - Funds
- ind_hip_fin_ult1 ==> Target column - Mortgage
- ind_plan_fin_ult1 ==> Target column - Pensions
- ind_pres_fin_ult1 ==> Target column - Loans
- ind_reca_fin_ult1 ==> Target column - Taxes
- ind_tjcr_fin_ult1 ==> Target column - Credit Card
- ind_valo_fin_ult1 ==> Target column - Securities
- ind_viv_fin_ult1 ==> Target column - Home Account
- ind_nomina_ult1 ==> Target column - Payroll
- ind_nom_pens_ult1 ==> NBC - Target column - Pensions
- ind_recibo_ult1 ==> Target column - Direct Debit

In [9]:
bank_total.dtypes

fecha_dato                object
ncodpers                   int32
ind_empleado              object
pais_residencia           object
sexo                      object
age                       object
fecha_alta                object
ind_nuevo                float32
antiguedad                object
indrel                   float32
ult_fec_cli_1t            object
indrel_1mes               object
tiprel_1mes               object
indresi                   object
indext                    object
conyuemp                  object
canal_entrada             object
indfall                   object
tipodom                  float32
cod_prov                 float32
nomprov                   object
ind_actividad_cliente    float32
renta                    float32
segmento                  object
ind_ahor_fin_ult1           int8
ind_aval_fin_ult1           int8
ind_cco_fin_ult1            int8
ind_cder_fin_ult1           int8
ind_cno_fin_ult1            int8
ind_ctju_fin_ult1           int8
ind_ctma_f

# <font color=blue>TRANSFORMATIONS</font>

#### <font color=blue>We can appreciate that:

* columns 'fecha_dato', 'fecha_alta' and 'ult_fec_cli_1t' are classified as object but are dates.
* column 'conyuemp' have null values, I'm going to consider them as 'N'.
* column 'age' have spaces on number fields and 27734 fields with NA. I'm going to convert all column to numbers and drop those NA values since it represent a 0.20% of all data and I'm going to follow "less is more" principle.
* column 'antiguedad' have NAs, spaces and wrong number values. I'm going to replace all and convert all column to numbers.
* column 'indrel_1mes' have values represented as float, integers... I'm going to standarize it. Additionally, this column haves 149,781 rows with null values, those rows represent a 1,10% of all data. However, inspecting test file I've found there are some nulls as well. So I have to keep those null values.
* column 'tiprel_1mes' have some nan values that represents less that 1,10% of all data. Same as before, I've found on test file some nan values, so I gonna keep them.
* column 'indresi'

</font>

In [27]:
def transformations(df):
    df['fecha_dato'] = pd.to_datetime(df['fecha_dato'], errors='coerce', infer_datetime_format=True, format="%Y-%m-%d")
    df['fecha_alta'] = pd.to_datetime(df['fecha_alta'], errors='coerce', infer_datetime_format=True, format="%Y-%m-%d")
    df['ult_fec_cli_1t'] = pd.to_datetime(bank_total['ult_fec_cli_1t'], errors='coerce', infer_datetime_format=True, format="%Y-%m-%d")
    df['conyuemp'].fillna('N', inplace=True)
    df['age'].replace(' ', '', inplace=True, regex=True)
    df['age'] = pd.to_numeric(df['age'], errors='ignore', downcast='integer')
    
    df['antiguedad'].replace(' NA', 0, inplace=True)
    df['antiguedad'].replace('NA', 0, inplace=True)
    df['antiguedad'].replace(' ', '', inplace=True, regex=True)
    df['antiguedad'].replace('-999999', 0, inplace=True, regex=True)
    df['antiguedad'] = pd.to_numeric(df['antiguedad'], errors='ignore', downcast='integer')
    df['indrel_1mes'].replace('1.0', '1', inplace=True)
    df['indrel_1mes'].replace('2.0', '2', inplace=True)
    df['indrel_1mes'].replace('3.0', '3', inplace=True)
    df['indrel_1mes'].replace('4.0', '4', inplace=True)
    df['indrel_1mes'].replace('1.', '1', inplace=True)
    df['indrel_1mes'].replace('2.', '2', inplace=True)
    df['indrel_1mes'].replace('3.', '3', inplace=True)
    df['indrel_1mes'].replace('4.', '4', inplace=True)


In [ ]:
def dropper(df):
    df = df[df.age != ' NA']
    
    

In [28]:
transformations(bank_total)

In [35]:
print(bank_total.shape)
bank_total.head()

(13647309, 48)


fecha_dato  ncodpers ind_empleado pais_residencia sexo  age fecha_alta  \
0 2015-01-28   1375586            N              ES    H   35 2015-01-12   
1 2015-01-28   1050611            N              ES    V   23 2012-08-10   
2 2015-01-28   1050612            N              ES    V   23 2012-08-10   
3 2015-01-28   1050613            N              ES    H   22 2012-08-10   
4 2015-01-28   1050614            N              ES    V   23 2012-08-10   

   ind_nuevo  antiguedad  indrel  ... ind_hip_fin_ult1 ind_plan_fin_ult1  \
0        0.0           6     1.0  ...                0                 0   
1        0.0          35     1.0  ...                0                 0   
2        0.0          35     1.0  ...                0                 0   
3        0.0          35     1.0  ...                0                 0   
4        0.0          35     1.0  ...                0                 0   

  ind_pres_fin_ult1 ind_reca_fin_ult1 ind_tjcr_fin_ult1 ind_valo_fin_ult1  \
0                 0                 0                 0                 0   
1                 0                 0                 0                 0   
2                 0                 0                 0                 0   
3                 0                 0                 0                 0   
4                 0                 0                 0                 0   

  ind_viv_fin_ult1 ind_nomina_ult1  ind_nom_pens_ult1  ind_recibo_ult1  
0                0             0.0                0.0                0  
1                0             0.0                0.0                0  
2                0             0.0                0.0                0  
3                0             0.0                0.0                0  
4                0             0.0                0.0                0  

[5 rows x 48 columns]

In [30]:
bank_total.dtypes

fecha_dato               datetime64[ns]
ncodpers                          int32
ind_empleado                     object
pais_residencia                  object
sexo                             object
age                               int16
fecha_alta               datetime64[ns]
ind_nuevo                       float32
antiguedad                        int16
indrel                          float32
ult_fec_cli_1t           datetime64[ns]
indrel_1mes                      object
tiprel_1mes                      object
indresi                          object
indext                           object
conyuemp                         object
canal_entrada                    object
indfall                          object
tipodom                         float32
cod_prov                        float32
nomprov                          object
ind_actividad_cliente           float32
renta                           float32
segmento                         object
ind_ahor_fin_ult1                  int8


In [80]:
bank_total['fecha_alta'].isna().sum()

27734

In [74]:
bank_total['age'].unique()

array([ 35,  23,  22,  24,  65,  28,  25,  26,  53,  27,  32,  37,  31,
        39,  63,  33,  55,  42,  58,  38,  50,  30,  45,  44,  36,  29,
        60,  57,  67,  47,   0,  34,  48,  46,  54,  84,  15,  12,   8,
         6,  83,  40,  77,  69,  52,  59,  43,  10,   9,  49,  41,  51,
        78,  16,  11,  73,  62,  66,  17,  68,  82,  95,  96,  56,  61,
        79,  72,  14,  19,  13,  86,  64,  20,  89,  71,   7,  70,  74,
        21,  18,  75,   4,  80,  81,   5,  76,  92,  93,  85,  91,  87,
        90,  94,  99,  98,  88,  97, 100, 101, 106, 103,   3,   2, 102,
       104, 111, 107, 109, 105, 112, 115, 110, 116, 108, 113, 126, 117,
       163, 127, 114, 164])

In [73]:
bank_total[bank_total['indresi'].isnull()]

fecha_dato  ncodpers ind_empleado pais_residencia sexo  age  \
261     2015-01-28   1050741          NaN             NaN  NaN    0   
1029    2015-01-28   1051017          NaN             NaN  NaN    0   
1063    2015-01-28   1051064          NaN             NaN  NaN    0   
1154    2015-01-28   1051387          NaN             NaN  NaN    0   
1779    2015-01-28   1048660          NaN             NaN  NaN    0   
1850    2015-01-28   1049775          NaN             NaN  NaN    0   
1867    2015-01-28   1049774          NaN             NaN  NaN    0   
1886    2015-01-28   1049838          NaN             NaN  NaN    0   
1922    2015-01-28   1049700          NaN             NaN  NaN    0   
2142    2015-01-28   1049406          NaN             NaN  NaN    0   
2418    2015-01-28   1049560          NaN             NaN  NaN    0   
2487    2015-01-28   1056017          NaN             NaN  NaN    0   
2989    2015-01-28   1055504          NaN             NaN  NaN    0   
3343    2015-01-28   1057040          NaN             NaN  NaN    0   
3411    2015-01-28   1056554          NaN             NaN  NaN    0   
3414    2015-01-28   1056525          NaN             NaN  NaN    0   
3550    2015-01-28   1056756          NaN             NaN  NaN    0   
3646    2015-01-28   1056643          NaN             NaN  NaN    0   
4150    2015-01-28   1052877          NaN             NaN  NaN    0   
4182    2015-01-28   1052940          NaN             NaN  NaN    0   
4281    2015-01-28   1054659          NaN             NaN  NaN    0   
4290    2015-01-28   1054673          NaN             NaN  NaN    0   
4362    2015-01-28   1054509          NaN             NaN  NaN    0   
4646    2015-01-28   1054037          NaN             NaN  NaN    0   
4775    2015-01-28   1054429          NaN             NaN  NaN    0   
4777    2015-01-28   1054432          NaN             NaN  NaN    0   
5288    2015-01-28   1043686          NaN             NaN  NaN    0   
5413    2015-01-28   1044036          NaN             NaN  NaN    0   
5503    2015-01-28   1045095          NaN             NaN  NaN    0   
5666    2015-01-28   1045269          NaN             NaN  NaN    0   
...            ...       ...          ...             ...  ...  ...   
3770661 2015-06-28    563245          NaN             NaN  NaN    0   
3770746 2015-06-28    563173          NaN             NaN  NaN    0   
3770881 2015-06-28    563449          NaN             NaN  NaN    0   
3771376 2015-06-28    565207          NaN             NaN  NaN    0   
3771380 2015-06-28    565160          NaN             NaN  NaN    0   
3771435 2015-06-28    569301          NaN             NaN  NaN    0   
3771523 2015-06-28    565649          NaN             NaN  NaN    0   
3771727 2015-06-28    544873          NaN             NaN  NaN    0   
3772008 2015-06-28    543501          NaN             NaN  NaN    0   
3772627 2015-06-28    545184          NaN             NaN  NaN    0   
3772783 2015-06-28    545692          NaN             NaN  NaN    0   
3772931 2015-06-28    540738          NaN             NaN  NaN    0   
3773459 2015-06-28    542881          NaN             NaN  NaN    0   
3773729 2015-06-28    541748          NaN             NaN  NaN    0   
3773767 2015-06-28    541866          NaN             NaN  NaN    0   
3774107 2015-06-28    552221          NaN             NaN  NaN    0   
3774329 2015-06-28    552715          NaN             NaN  NaN    0   
3774635 2015-06-28    551847          NaN             NaN  NaN    0   
3774795 2015-06-28    554453          NaN             NaN  NaN    0   
3774843 2015-06-28    554887          NaN             NaN  NaN    0   
3774902 2015-06-28    554521          NaN             NaN  NaN    0   
3775205 2015-06-28    553542          NaN             NaN  NaN    0   
3775245 2015-06-28    553680          NaN             NaN  NaN    0   
3775860 2015-06-28    547784          NaN             NaN  NaN    0   
3775909 2015-06-28    550339    

In [ ]:
df.loc[df.age < 18,"age"]  = df.loc[(df.age >= 18) & (df.age <= 30),"age"].mean(skipna=True)
df.loc[df.age > 100,"age"] = df.loc[(df.age >= 30) & (df.age <= 100),"age"].mean(skipna=True)
df["age"].fillna(df["age"].mean(),inplace=True)
df["age"] = df["age"].astype(int)

In [40]:
test = pd.read_csv('./data/test_ver2.csv')

/home/alfonso/miniconda3/envs/kaggle/lib/python3.7/site-packages/IPython/core/interactiveshell.py:3057: DtypeWarning: Columns (15) have mixed types. Specify dtype option on import or set low_memory=False.
  interactivity=interactivity, compiler=compiler, result=result)


In [43]:
test.isnull().sum()

fecha_dato                    0
ncodpers                      0
ind_empleado                  0
pais_residencia               0
sexo                          5
age                           0
fecha_alta                    0
ind_nuevo                     0
antiguedad                    0
indrel                        0
ult_fec_cli_1t           927932
indrel_1mes                  23
tiprel_1mes                  23
indresi                       0
indext                        0
conyuemp                 929511
canal_entrada              2081
indfall                       0
tipodom                       0
cod_prov                   3996
nomprov                    3996
ind_actividad_cliente         0
renta                         0
segmento                   2248
dtype: int64

In [48]:
test.shape

(929615, 24)

In [49]:
test.dtypes

fecha_dato                object
ncodpers                   int64
ind_empleado              object
pais_residencia           object
sexo                      object
age                        int64
fecha_alta                object
ind_nuevo                  int64
antiguedad                 int64
indrel                     int64
ult_fec_cli_1t            object
indrel_1mes              float64
tiprel_1mes               object
indresi                   object
indext                    object
conyuemp                  object
canal_entrada             object
indfall                   object
tipodom                    int64
cod_prov                 float64
nomprov                   object
ind_actividad_cliente      int64
renta                     object
segmento                  object
dtype: object

In [76]:
test['ind_nuevo'].unique()

array([0, 1])